File: python/x402/mechanisms/evm/erc6492.py
"""ERC-6492 signature parsing utilities."""

try:
    from eth_abi import decode
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .constants import ERC6492_MAGIC_VALUE
from .types import ERC6492SignatureData


def is_erc6492_signature(signature: bytes) -> bool:
    """Check if signature has ERC-6492 magic suffix.

    ERC-6492 signatures end with 32-byte magic value.

    Args:
        signature: Signature bytes.

    Returns:
        True if signature is ERC-6492 wrapped.
    """
    if len(signature) < 32:
        return False
    return signature[-32:] == ERC6492_MAGIC_VALUE


def parse_erc6492_signature(signature: bytes) -> ERC6492SignatureData:
    """Parse ERC-6492 wrapped signature.

    ERC-6492 Format:
        abi.encode((address factory, bytes factoryCalldata, bytes signature)) + magicBytes

    If not ERC-6492, returns original signature as inner_signature
    with empty factory/calldata.

    Args:
        signature: Signature bytes (may or may not be ERC-6492).

    Returns:
        Parsed signature data.

    Raises:
        ValueError: If ERC-6492 format is invalid.
    """
    if not is_erc6492_signature(signature):
        return ERC6492SignatureData(
            factory=bytes(20),
            factory_calldata=b"",
            inner_signature=signature,
        )

    # Strip magic value
    payload = signature[:-32]

    # Decode ABI: (address, bytes, bytes)
    try:
        factory, factory_calldata, inner_signature = decode(
            ["address", "bytes", "bytes"],
            payload,
        )

        # Convert factory address to bytes
        if isinstance(factory, str):
            factory_bytes = bytes.fromhex(factory.removeprefix("0x"))
        else:
            factory_bytes = factory

        return ERC6492SignatureData(
            factory=factory_bytes,
            factory_calldata=factory_calldata,
            inner_signature=inner_signature,
        )
    except Exception as e:
        raise ValueError(f"Invalid ERC-6492 signature format: {e}") from e


def is_eoa_signature(sig_data: ERC6492SignatureData) -> bool:
    """Check if signature is from an EOA (65 bytes, no factory).

    Args:
        sig_data: Parsed signature data.

    Returns:
        True if EOA signature.
    """
    zero_factory = bytes(20)
    return len(sig_data.inner_signature) == 65 and sig_data.factory == zero_factory


def has_deployment_info(sig_data: ERC6492SignatureData) -> bool:
    """Check if signature has smart wallet deployment info.

    Args:
        sig_data: Parsed signature data.

    Returns:
        True if has factory and calldata for deployment.
    """
    zero_factory = bytes(20)
    return sig_data.factory != zero_factory and len(sig_data.factory_calldata) > 0

File: go/mechanisms/evm/erc6492.go
package evm

import (
	"bytes"
	"fmt"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
)

// erc6492MagicBytes is the 32-byte magic value suffix for ERC-6492 signatures
// This is bytes32(uint256(keccak256("erc6492.invalid.signature")) - 1)
var erc6492MagicBytes = common.Hex2Bytes(
	"6492649264926492649264926492649264926492649264926492649264926492",
)

// IsERC6492Signature checks if a signature has the ERC-6492 magic suffix
//
// ERC-6492 signatures are wrapped signatures for counterfactual smart contract accounts.
// They end with a specific 32-byte magic value to distinguish them from regular signatures.
//
// Args:
//
//	sig: The signature bytes to check
//
// Returns:
//
//	true if the signature ends with the ERC-6492 magic value
func IsERC6492Signature(sig []byte) bool {
	if len(sig) < 32 {
		return false
	}
	return bytes.Equal(sig[len(sig)-32:], erc6492MagicBytes)
}

// ParseERC6492Signature unwraps an ERC-6492 signature to extract its components
//
// ERC-6492 Format:
//
//	abi.encode((address factory, bytes factoryCalldata, bytes signature)) + magicBytes
//
// If the signature is not ERC-6492 format, it returns the original signature
// as the InnerSignature with empty Factory and FactoryCalldata.
//
// Args:
//
//	sig: The signature bytes (may or may not be ERC-6492 wrapped)
//
// Returns:
//
//	ERC6492SignatureData containing the parsed components
//	error if the ERC-6492 format is invalid
func ParseERC6492Signature(sig []byte) (*ERC6492SignatureData, error) {
	// If not ERC-6492, return original signature
	if !IsERC6492Signature(sig) {
		return &ERC6492SignatureData{
			InnerSignature: sig,
		}, nil
	}

	// Strip magic value
	payload := sig[:len(sig)-32]

	// Define ABI types for (address, bytes, bytes)
	addressTy, err := abi.NewType("address", "", nil)
	if err != nil {
		return nil, err
	}
	bytesTy, err := abi.NewType("bytes", "", nil)
	if err != nil {
		return nil, err
	}

	arguments := abi.Arguments{
		{Type: addressTy}, // factory
		{Type: bytesTy},   // factoryCalldata
		{Type: bytesTy},   // originalSignature
	}

	// Unpack the ABI-encoded data
	unpacked, err := arguments.Unpack(payload)
	if err != nil {
		return nil, err
	}

	if len(unpacked) != 3 {
		return nil, fmt.Errorf("invalid ERC-6492 signature: expected 3 fields, got %d", len(unpacked))
	}

	factory, ok := unpacked[0].(common.Address)
	if !ok {
		return nil, fmt.Errorf("invalid ERC-6492 signature: factory is not an address")
	}

	factoryCalldata, ok := unpacked[1].([]byte)
	if !ok {
		return nil, fmt.Errorf("invalid ERC-6492 signature: factoryCalldata is not bytes")
	}

	innerSignature, ok := unpacked[2].([]byte)
	if !ok {
		return nil, fmt.Errorf("invalid ERC-6492 signature: innerSignature is not bytes")
	}

	// Convert factory address to [20]byte
	var factoryBytes [20]byte
	copy(factoryBytes[:], factory.Bytes())

	return &ERC6492SignatureData{
		Factory:         factoryBytes,
		FactoryCalldata: factoryCalldata,
		InnerSignature:  innerSignature,
	}, nil
}
